nexus/social_media/services.py

```diff
@@ -0,0 +1,173 @@
+            },
+        },
+        "visibility": {
+            "com.linkedin.ugc.MemberNetworkVisibility": "PUBLIC"
+        },
+    }
+
+    if post.image:
+        response_upload_image, asset = upload_image_to_linkedin(author,
+                                                                headers,
+                                                                post,
+                                                                api_url_base,
+                                                                linkedin)
+
+        if response_upload_image.status_code == status.HTTP_201_CREATED:
+            shareContent = post_data.get('specificContent')\
+                                    .get('com.linkedin.ugc.ShareContent')
+            image_media = [{
+                "status": "READY",
+                "description": {
+                    "text": ""
+                },
+                "media": asset,
+                "title": {
+                    "text": ""
+                }
+            }]
+            shareContent.update(shareMediaCategory='IMAGE',
+                                media=image_media)
+            image_response = requests.post(api_url_ugc, headers=headers,
+                                           json=post_data)
+            return image_response
+        else:
+            return response_upload_image
+    elif post.text:
+        response = requests.post(api_url_ugc, headers=headers, json=post_data)
+        return response
+
+
+def appropriate_response_action(response):
+    """Function to perform appropriate action based on the response given.
+
+    :param response: A valid HTTP response to perform suitable action.
+
+    :raises WrongArguments: Exception raised when either a 400 or 404 response is provided.
+
+    :raises NotAuthenticated: Exception raised when a 401 response is provided.
+
+    :raises PermissionDenied: Exception raised when a 403 response is provided.
+
+    :raises NotSupported: Exception raised when a 405 response is provided.
+
+    """
+    status_code = response.status_code
+    if status_code != status.HTTP_201_CREATED:
+        json_response = response.json()
+        error = f"LinkedIn error: {json_response['message']}"
+        if status_code == status.HTTP_400_BAD_REQUEST:
+            raise exc.WrongArguments(error)
+        elif status_code == status.HTTP_401_UNAUTHORIZED:
+            raise exc.NotAuthenticated(error)
+        elif status_code == status.HTTP_403_FORBIDDEN:
+            raise exc.PermissionDenied(error)
+        elif status_code == status.HTTP_404_NOT_FOUND:
+            raise exc.WrongArguments(error)
+        elif status_code == status.HTTP_405_METHOD_NOT_ALLOWED:
+            raise exc.NotSupported(error)
+
+
+def publish_post_service():
+    """Function to publish social media post"""
+    posts = Post.objects.filter(
+        is_approved=True, is_posted=False, scheduled_time__lte=timezone.now()
+    )
+    for post in posts:
```

Comment:
Why not update all the queryset in bulk and then ensure that it gets posted asynchronously. Once it is posted, each task can make the `is_posted` set to `True`.

Reply:
Isn't all the posts being updated in bulk? I am first updating all the posts then posting them one by one. Since, this service is being called by the celery task, the posts will be posted asynchronously.
Or Do you mean like only saving `posted_time` in bulk and updating `is_posted` to `True` as each post get posted?

Reply:
@gutsytechster You may update posts in bulk using `posts.update()` instead of using a loop.